doc/source/ray-tracing.rst

```diff
@@ -0,0 +1,101 @@
+Tracing
+=======
+To help debug and monitor Ray applications, Ray integrates with OpenTelemetry to make it easy to export traces to external tracing stacks such as Jaeger. 
+
+
+.. note::
+
+    Tracing is currently an experimental feature and under active development. APIs are subject to change.
+
+Getting Started
+---------------
+First, install OpenTelemetry.
+.. code-block:: bash
+
+    pip install opentelemetry-api==1.0.0rc1
+    pip install opentelemetry-sdk==1.0.0rc1
+
+To enable tracing, you must provide a tracing startup hook with a function that will set up the Tracer Provider, Remote Span Processors, and Additional Instruments. The tracing startup hook is expected to be a function that will be called with no args or kwargs.
+Tracer Provider
+~~~~~~~~~~~~~~~~
+This configures how to collect traces. View the TracerProvider API `here <https://open-telemetry.github.io/opentelemetry-python/sdk/trace.html#opentelemetry.sdk.trace.TracerProvider>`__.
+
+Remote Span Processors
+~~~~~~~~~~~~~~~~~~~~~~
+This configures where to export traces to. View the SpanProcessor API `here <https://open-telemetry.github.io/opentelemetry-python/sdk/trace.html#opentelemetry.sdk.trace.SpanProcessor>`__.
+
+Users who want to experiment with tracing can configure their remote span processors to export spans to a local JSON file. Serious users developing locally can push their traces to Jaeger containers via the `Jaeger exporter <https://open-telemetry.github.io/opentelemetry-python/exporter/jaeger/jaeger.html>`_.
+
+
+Additional Instruments
+~~~~~~~~~~~~~~~~~~~~~~
+If you are using a library that has built-in tracing support, the ``setup_tracing`` function you provide should also patch those libraries. You can find more documentation for the instrumentation of these libraries `here <https://github.com/open-telemetry/opentelemetry-python-contrib/tree/main/instrumentation>`_.
+
+Below is an example tracing startup hook that sets up the default tracing provider, exports spans to files in ``/tmp/spans``, and does not have any additional instruments.
+
+.. code-block:: python
+
+  import ray
+  import os
+  from opentelemetry import trace
+  from opentelemetry.sdk.trace import TracerProvider
+  from opentelemetry.sdk.trace.export import (
+      ConsoleSpanExporter,
+      SimpleExportSpanProcessor,
+  )
+  
+  
+  def setup_tracing() -> None:
```

Comment:
This is a question and not a suggestion, but I'm still having trouble wrapping my head around what this setup function is supposed to do.

There's some gap in my understanding, but I'm imagining that Ray is generating events (e.g., tuples like `('event-name', start_timestamp, end_timestamp)` or something) and we want those events to get sent somewhere (e.g., a file or a service or something).

So I would imagine that the information the user needs to provide is
1. Tell Ray where to send the events
2. Optionally what format to put the events in (but maybe everything is standardized and this is unnecessary)

Given that, I would expect the configuration to either look like
1. The address of the place to write the events.
2. A function that takes in an event, formats it, and writes it to the correct place.

The code here doesn't really seem to fit either of those descriptions.

Reply:
Ah I see what you are saying! Tracing can be thought of as a different process than metrics (where metrics would be more similar to gathering information outputted by Ray).  You can think of tracing as simply wrapping different events and gathering all the necessary information of timestamps, event names, etc. automatically. So if you wrap a code sample with `with tracer.start_as_current_span("foo"):`, the name of the span would be `foo` and it would automatically track the timestamps and such! Does that make sense? Rather than collecting events after they have occurred, tracing is gathering the information *as the events occur.*

Reply:
Additionally, users are not telling Ray where to send the events, they are telling OpenTelemetry where to send it. Ray just push events to OpenTelemetry functions and those functions will send out the events to proper destination. 

This segment encodes "where to send it": 
```
trace.get_tracer_provider().add_span_processor(
          SimpleExportSpanProcessor(
              ConsoleSpanExporter(
                  out=open(f"/tmp/spans/{os.getpid()}.json", "a")
                  )
          )
      )
```

Reply:
Yeah the interesting point here is that there's nothing specific to tracing about the interface; we discussed offline and it seems to make sense to turn this into a general startup hook for worker processes that could be more widely useful.